File: search-race/mc.cpp
// https://www.codingame.com/ide/puzzle/search-race
#pragma GCC optimize("Ofast")
#pragma GCC optimize("inline")
#pragma GCC optimize("omit-frame-pointer")
#pragma GCC optimize("unroll-loops")
#include "math.h"
#include <iostream>
#include <vector>

// TODO: you shell experiment with different DEPTH values
#define DEPTH 1

using namespace std;

struct Point {
    int x, y;
};

inline int rnd(int b) {
    return rand() % b;
}

inline int rnd(int a, int b) {
    return a + rnd(b - a + 1);
}

/// Normalize angle to be in the range -180..180
int norm_angle(int a) {
    a = a % 360;
    if (a < -180) return a + 360;
    if (a > 180) return a - 360;
    return a;
}

struct Move {
    int x, y, thrust;

    Move(int x = 0, int y = 0, int thrust = 0)
        : x(x), y(y), thrust(thrust) {}

    void print() const {
        cout << x << " " << y << " " << thrust << endl;
    }
};

struct State {
    const vector<Point>& checkpoints; // immutable
    int checkpoint_index;
    double x, y, vx, vy, angle;

    State(const vector<Point>& checkpoints, int checkpoint_index, int x, int y, int vx, int vy, int angle)
        : checkpoints(checkpoints), checkpoint_index(checkpoint_index), x(x), y(y), vx(vx), vy(vy), angle(angle) {}

    State(const State& other)
        : checkpoints(other.checkpoints), checkpoint_index(other.checkpoint_index), x(other.x), y(other.y), vx(other.vx), vy(other.vy), angle(other.angle) {}

    void print() const {
        cerr << "State(" 
            << "checkpoint_index=" << checkpoint_index 
            << ", x=" << x << ", y=" << y
            << ", vx=" << vx << ", vy=" << vy 
            << ", angle=" << angle << ")" << endl;
    }

    Point get_checkpoint(int index = 0) const {
        return checkpoints[(checkpoint_index + index) % checkpoints.size()];
    }

    void simulate(const Move& move) {
        double desired_angle = 180.0 * atan2(move.y - y, move.x - x) / M_PI;
        int da = norm_angle(desired_angle - angle);
        da = max(-18, min(18, da));
        angle = angle + da;
        vx += move.thrust * cos(angle * M_PI / 180.0);
        vy += move.thrust * sin(angle * M_PI / 180.0);
        x = static_cast<int>(x + vx);
        y = static_cast<int>(y + vy);
        vx = static_cast<int>(0.85 * vx);
        vy = static_cast<int>(0.85 * vy);
        angle = static_cast<int>(round(angle)) % 360;
        Point cp = get_checkpoint();
        int dx = x - cp.x, dy = y - cp.y;
        if (dx * dx + dy * dy < 590 * 590) {
            checkpoint_index++;
        }
    }
};

double estimate(const State& state){
    // TODO: implement your own estimation function
    // Calculates the score of the state.
    // The higher the checkpoint_index, the better.
    // The closer to the next checkpoint, the better.
    // ... [your ideas here] ...
    return 0;
}

inline Move create_random_move(){
    // TODO: Create random Move instance:
    return Move(5000, 5000, 200);
} 

/// Random search with a time limit
void random_search(State& state, Move best_moves[], clock_t end_time) {
    double best_score = -std::numeric_limits<double>::infinity();
    int simulations_count = 0;
    while (clock()  < end_time) {
        // TODO: Finish an implementation of the Monte Carlo algorithm a.k.a random search
        // While still have time, create a new random sequence of moves (use create_random_move).
        // Simulate this sequence of moves (use state.simulate(move)).
        // Estimate the final state after simulation and store the best sequence. (use estimate(state))
        // Store the best move sequence in best_moves[]

        simulations_count++;
    }
    cerr << best_score << " sims: " << simulations_count << endl;
}

vector<Point> read_checkpoints() {
    int n;
    cin >> n;
    vector<Point> checkpoints(n);
    for (int i = 0; i < n; ++i)
        cin >> checkpoints[i].x >> checkpoints[i].y;
    return checkpoints;
}

int main() {
    vector<Point> checkpoints = read_checkpoints();

    Move best_moves[DEPTH];
    while (true) {
        int checkpoint_index, x, y, vx, vy, angle;
        cin >> checkpoint_index >> x >> y >> vx >> vy >> angle; // input current state

        /// Actual time per move is 50ms, but to avoid random timeouts we cut it in half
        clock_t end_time = clock() + 0.025 * CLOCKS_PER_SEC;
        State state(checkpoints, checkpoint_index, x, y, vx, vy, angle);
        random_search(state, best_moves, end_time);

        best_moves[0].print(); // output our next move!
    }
    return 0;
}

File: search-race/mc.py
import random
import sys
import math
import time


def norm_angle(a):
    a = a % 360  # a in 0..360
    if a > 180:
        a -= 360  # a in -180..180
    return a


class Move:
    def __init__(self, x, y, thrust, message=""):
        self.x, self.y, self.thrust, self.message = x, y, thrust, message

    def __str__(self):
        return f'{self.x} {self.y} {self.thrust} {self.message}'


class State:
    def __init__(self, checkpoints, checkpoint_index, x, y, vx, vy, angle):
        self.checkpoints = checkpoints
        self.checkpoint_index = checkpoint_index
        self.x = x
        self.y = y
        self.vx = vx
        self.vy = vy
        self.angle = angle


    def __str__(self):
        return f'State(checkpoints, {self.checkpoint_index}, {self.x}, {self.y}, {self.vx}, {self.vy}, {self.angle})'


    def copy(self):
        return State(self.checkpoints, self.checkpoint_index, self.x, self.y, self.vx, self.vy, self.angle)


    def cur_checkpoint(self):
        return self.checkpoints[self.checkpoint_index % len(self.checkpoints)]


    def simulate(self, move: Move):
        desired_angle = 180 * math.atan2(move.y - self.y, move.x - self.x) / math.pi
        da = norm_angle(desired_angle - self.angle)
        da = max(-18, min(18, da))
        self.angle = self.angle + da
        self.vx += move.thrust * math.cos(self.angle * math.pi / 180)
        self.vy += move.thrust * math.sin(self.angle * math.pi / 180)
        self.x = int(self.x + self.vx)
        self.y = int(self.y + self.vy)
        self.vx = int(0.85 * self.vx)
        self.vy = int(0.85 * self.vy)
        self.angle = round(self.angle) % 360

        # simplified checkpoint-collision check
        xc, yc = self.cur_checkpoint()
        dx, dy = self.x - xc, self.y - yc
        if dx * dx + dy * dy < 590 * 590:
            self.checkpoint_index += 1


def estimate(state):
    """
    Calculates the score of the state.
    The higher the checkpoint_index, the better.
    The closer to the next checkpoint, the better.
    ... [your ideas here] ...
    """
    # TODO 1
    pass


def create_random_moves(depth):
    """
    Returns an array of depth random Move objects.
    ... (or not so random) ...
    """
    # TODO 2
    pass


def random_search(state, depth):
    """
    Implementation of the Monte Carlo algorithm a.k.a random search
    
    While still have time, create a new random sequence of moves (use create_random_moves).
    Simulate this sequence of moves (use state.simulate(move)).
    Evaluate the final state after simulation and store the best sequence.

    When time (50ms) runs out, returns the best sequence of moves.
    
    To track time, use the time.time() function that returns the current time in seconds.
    """
    best_moves = []
    best_score = -math.inf
    simulations_count = 0  # number of random solutions that were considered
    
    # TODO 3: random search implementation here

    # debug output: best found score; number of estimated sequences; best sequence of moves
    print(best_score, simulations_count, list(map(str, best_moves)), file=sys.stderr)
    return best_moves


def read_checkpoints():
    n = int(input())
    checkpoints = []
    for i in range(n):
        x, y = [int(j) for j in input().split()]
        checkpoints.append((x, y))
    return checkpoints


def main():
    checkpoints = read_checkpoints()
    print(f'checkpoints={checkpoints}', file=sys.stderr)
    best_moves = None
    while True:
        state = State(checkpoints, *list(map(int, input().split())))
        print(state, file=sys.stderr)
        
        # TODO 4: experiment with different depths
        best_moves = random_search(state, depth=1)
        best_move = best_moves[0]
        print(best_move)

if __name__ == '__main__':
    main()